src/main/kotlin/io/github/mojira/arisa/registry/ModuleRegistry.kt

```diff
@@ -0,0 +1,87 @@
+package io.github.mojira.arisa.registry
+
+import arrow.core.Either
+import arrow.core.left
+import com.uchuhimo.konf.Config
+import io.github.mojira.arisa.ModuleExecutor
+import io.github.mojira.arisa.domain.Issue
+import io.github.mojira.arisa.infrastructure.config.Arisa
+import io.github.mojira.arisa.infrastructure.config.Arisa.Modules.ModuleConfigSpec
+import io.github.mojira.arisa.modules.FailedModuleResponse
+import io.github.mojira.arisa.modules.Module
+import io.github.mojira.arisa.modules.ModuleError
+import io.github.mojira.arisa.modules.ModuleResponse
+import java.time.Instant
+
+// All defined module registries
+val getModuleRegistries = { config: Config ->
+    listOf(
+        InstantModuleRegistry(config),
+        DelayedModuleRegistry(config),
+        LazyModuleRegistry(config)
+    )
+}
+
+abstract class ModuleRegistry(private val config: Config) {
+    data class Entry(
+        val name: String,
+        val config: ModuleConfigSpec,
+        val execute: (issue: Issue, lastRun: Instant) -> Pair<String, Either<ModuleError, ModuleResponse>>,
+        val executor: ModuleExecutor
+    )
+
+    private val modules = mutableListOf<Entry>()
+
+    protected abstract val getJql: (Instant) -> String
+
+    fun getAllModules(): List<Entry> = modules
+
+    fun getEnabledModules(): List<Entry> = modules.filter(::isModuleEnabled)
+
+    private fun isModuleEnabled(module: Entry) =
+        // If arisa.debug.enabledModules is defined, return whether that module is in that list.
+        // If it's not defined, return whether this module is enabled in the module config.
+        config[Arisa.Debug.enabledModules]?.contains(module.name) ?: config[module.config.enabled]
+
+    protected fun register(
+        moduleConfig: ModuleConfigSpec,
+        module: Module
+    ) {
+        val moduleName = moduleConfig::class.simpleName!!
+
+        modules.add(
+            Entry(
+                moduleName,
+                moduleConfig,
+                getModuleResult(moduleName, module),
+                ModuleExecutor(config, moduleConfig)
+            )
+        )
+    }
+
+    private fun getModuleResult(moduleName: String, module: Module) = { issue: Issue, lastRun: Instant ->
+        moduleName to tryExecuteModule { module(issue, lastRun) }
+    }
+
+    private fun getJqlWithDebug(lastRun: Instant): String {
+        val debugWhitelist = config[Arisa.Debug.ticketWhitelist]
+
+        return if (debugWhitelist == null) getJql(lastRun)
+        else "key IN (${debugWhitelist.joinToString()}) AND (${getJql(lastRun)})"
```

Comment:
Maybe store `getJql(lastRun)` in a variable and then use its value instead to make it more obvious that this is always considered (instead of having it duplicated as it is currently the case).